### DISCUSSION TITLE
How can I send script output from cron?

### DISCUSSION THREAD

**Discussion by hugalafutro** [healthchecks/healthchecks#719] [Q&A] [answered]
Hi,
I am trying to send the script output of https://github.com/framps/raspiBackup via a script named `myraspibkp.sh`:
```
#!/bin/bash
curl -fsS --retry 3 https://[REDACTED]/ping/[REDACTED]/start
m=$(/usr/local/bin/raspiBackup)
curl -fsS --retry 3 -m10 --data-raw "$m" https://[REDACTED]/ping/[REDACTED]/$?
```
When I run it manually I get the desired output like so:
![image](https://user-images.githubusercontent.com/30209689/197332310-4d7319f1-16a4-426d-8141-25f0bc9eb8cf.png)

When it runs via cron line `11   4  *   *   6     bash /home/pi/myraspibkp.sh 2>&1` though I get nothing, as in the whole Request Body section just isn't there

Is there any way to get the output when running from cron?

Thanks for any info.

**Comment by cuu508** [comment, marked as answer]
When it runs from cron,

* do you see the "started" and "ok" events in the event log?
* does it look like raspiBackup did in fact run? i.e., did it produce new backups and logfiles?

As an experiment, I just did the following on RPi 400 running Raspbian: 

Created `/home/user/experiment.sh` and made it executable:

```
#!/bin/bash
curl -fsS --retry 3 https://hc-ping.com/my-uuid-here/start
m=$(/usr/games/cowsay hello)
curl -fsS --retry 3 -m10 --data-raw "$m" https://hc-ping.com/my-uuid-here/$?
```

Installed cowsay (`sudo apt install cowsay`) and created a cron entry for it:

```
* * * * * bash /home/user/experiment.sh 2>&1
```

I do see the success pings coming in every minute, and they do contain the request body with cowsay output. 

Perhaps you can do something similar, to see if the problem is specific to raspiBackup, or it's something else.

**Reply by hugalafutro** [reply]
Everything works, the log+backup gets created just the output doesn't get sent, but I tried with the cowsay and that works so I guess I am taking this over to raspiBackup's github.

Thanks for helping me narrowing it down!

**Comment by hugalafutro** [comment]
I got it working like this (per discussion with author stdout is disabled when running headless)
```
#!/bin/bash
if [ "$(id -u)" -ne 0 ]; then
        echo 'Run me as root :)' >&2
        exit 1
fi
curl -fsS --retry 3 https://[REDACTED]/ping/[REDACTED]/start
/usr/local/bin/raspiBackup -L /tmp/raspiBackup
m=$(cat /tmp/raspiBackup.msg)
curl -fsS --retry 3 -m10 --data-raw "$m" https://[REDACTED]/ping/[REDACTED]/$?
```

But if I understand things correctly I am now receiving the exit code of the `cat` command which would be not failing even if the previous command failed. Is there any way I could do some bash magic to get it working like it is, but instead return the exit code of the backup command (hope that makes sense)?
Thanks for any suggestion!

**Reply by cuu508** [reply]
After running raspiBackup, store the exit code in a variable so it does not get overwritten when running cat. A quick experiment to illustrate:

```
cat /this-does-not-exist
RESULT=$?
echo hey
echo the exit code of the first command is $RESULT
```

**Reply by cuu508** [reply]
This also seems to work:

```bash
# Prepare a text file to submit:
echo hello world > /tmp/message
# Run a command that will return with a non-zero exit code:
cat /this-does-not-exist
# Ping healthchecks, tell curl to read request body from file:
curl -fsS --retry 3 -m10 --data-binary "@/tmp/message" https://hc-ping.com/my-uuid-here/$?
```
